HW2/P2/parallel_vector.pyx

```diff
@@ -54,39 +54,55 @@ cpdef move_data_serial(np.int32_t[:] counts,
                        np.int32_t[:] src,
                        np.int32_t[:] dest,
                        int repeat):
-   cdef:
-       int idx, r
+    cdef:
+        int idx, r
 
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+    assert src.size == dest.size, "Sizes of src and dest arrays must match"
+    with nogil:
+        for r in range(repeat):
+            for idx in range(src.shape[0]):
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
+    cdef:
+        int idx, r
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
 
    ##########
    # Your code here
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+    for r in range(repeat):
+        for idx in prange(src.shape[0],nogil=True,num_threads = 4,schedule=dynamic):
+            if dest[idx] < src[idx]: # lock each element individually, sort to prevent deadlock
+                omp_set_lock(&(locks[dest[idx]]))
+                omp_set_lock(&(locks[src[idx]]))
+            elif dest[idx] > src[idx]:
+                omp_set_lock(&(locks[src[idx]]))
+                omp_set_lock(&(locks[dest[idx]]))
+            else:
+                omp_set_lock(&(locks[src[idx]])) # prevent double locking here
+                           
+            if counts[src[idx]] > 0:
+                counts[dest[idx]] += 1
+                counts[src[idx]] -= 1
+
+            if dest[idx] < src[idx]:
+                omp_unset_lock(&(locks[dest[idx]]))
+                omp_unset_lock(&(locks[src[idx]]))
+            elif dest[idx] > src[idx]:
+                omp_unset_lock(&(locks[src[idx]]))
+                omp_unset_lock(&(locks[dest[idx]]))
+            else:
+                omp_unset_lock(&(locks[src[idx]]))
+    free_N_locks(counts.shape[0], locks)
```

Comment:
also, usually, you want to do a check again after acquiring the lock, since the condition might have changed between checking the condition and acquiring the lock